### ISSUE TITLE
Refactor out field type switch statements

### ISSUE THREAD

**Issue by JedWatson** [keystonejs/keystone#1150]
We have a lot of switch statements throughout Keystone where certain field types have special behaviours, e.g.

```
switch(field.type){
  case "markdown": //do something
}
```

These are a major blocker for being able to implement custom field types, and contain field-specific logic in the wrong places.

We need to identify where and why these exist, and replace them with hooks that FieldTypes can attach to as required.

**Comment by creynders** [comment]
I'll be taking a shot at this. I've already encountered a few of those spots. I'll list them out here along the way.

**Comment by creynders** [comment]
#### definitely

* https://github.com/keystonejs/keystone/blob/master/lib/list.js#L904
* https://github.com/keystonejs/keystone/blob/master/lib/updateHandler.js#L221
* https://github.com/keystonejs/keystone/blob/master/templates/mixins/columns.jade#L1
* https://github.com/keystonejs/keystone/blob/master/templates/views/list.jade#L234
* https://github.com/keystonejs/keystone/blob/master/lib/content/page.js#L99

#### debatable

* https://github.com/keystonejs/keystone/blob/master/admin/src/components/CreateForm.js#L94
* https://github.com/keystonejs/keystone/blob/master/lib/list.js#L859
* https://github.com/keystonejs/keystone/blob/master/lib/core/createItems.js#L109
* https://github.com/keystonejs/keystone/blob/master/routes/download/list.js#L16
* https://github.com/keystonejs/keystone/blob/master/routes/download/list.js#L30
* https://github.com/keystonejs/keystone/blob/master/routes/download/list.js#L32
* https://github.com/keystonejs/keystone/blob/master/routes/views/item.js#L58
* https://github.com/keystonejs/keystone/blob/master/routes/views/list.js#L44
* https://github.com/keystonejs/keystone/blob/master/templates/views/list.jade#L94

This was only after a first, shallow search.

**Comment by creynders** [comment]
So, conceptually this translates to field-specific

* mongoose query parameters for searching
* error handling upon document updates
* column rendering on list pages
* filter rendering on list pages
* not entirely sure what lib/content/page is used for ATM

**Comment by geloescht** [comment]
Assessed this today for current 0.4 master 774d64ea3c1c9ec :
lib/updateHandler.js has its own issue #1194
lib/list/getSearchFilters.js  is deprecated according to the comment at the top of the file. Maybe a deprecation warning should be printed upon the first invocation.
**lib/list/addSearchToQuery.js** has some special casing of name fields
lib/list/field.js, lib/list/expandColumns.js, lib/core/createItems.js, lib/list/getCSV.js, lib/list/getData.js, admin/server/api/item/get.js all have some special casing of relationship fields. These are not an issue IMO, as relationship fields *are* special.
**admin/server/api/download.js** has special casing of boolean and relationship fields

Will try to come up solution for the files marked in bold, then I believe this can be closed.

**Comment by wmertens** [comment]
I do believe the only place with cases now is `updateHandler`, which will be gone soon :)

I suppose this can be closed?

**Comment by JedWatson** [comment]
This is good to close, it'll be resolved by #1150
